doc/source/whatsnew/v0.25.0.rst

```diff
@@ -63,6 +63,47 @@ is respected in indexing. (:issue:`24076`, :issue:`16785`)
     df = pd.DataFrame([0], index=pd.DatetimeIndex(['2019-01-01'], tz='US/Pacific'))
     df['2019-01-01 12:00:00+04:00':'2019-01-01 13:00:00+04:00']
 
+.. _whatsnew_0250.api_breaking.groupby_apply_first_group_once:
+
+GroupBy.apply on ``DataFrame`` evaluates first group only once
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The implementation of :meth:`DataFrameGroupBy.apply() <pandas.core.groupby.DataFrameGroupBy.apply>`
+previously evaluated the supplied function consistently twice on the first group
+to infer if it is safe to use a fast code path. Particularly for functions with
```

Comment:
Hmm I'm somewhat hesitant to market "side effects" as being fixed because I'm not sure we generally can make guarantees about mutations that occur during a groupby. Is it possible to leave this out and maybe use a relatively safe example like printing the groups instead?

Reply:
I agree with @WillAyd  here, showing the printing is enough (and a bit more clear).